src/main/java/com/treasuredata/android/cdp/SegmentResultsOfJSON.java

```diff
@@ -0,0 +1,87 @@
+package com.treasuredata.android.cdp;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.json.JSONTokener;
+
+import java.util.ArrayList;
+import java.util.List;
+
+abstract class JSONLookupResult extends LookupResult {
+    final int statusCode;
+
+    JSONLookupResult(int statusCode) {
+        this.statusCode = statusCode;
+    }
+
+    /**
+     * @throws JSONException            if the provided body is not json
+     * @throws IllegalArgumentException if the parsed json is neither object or array, or has unexpected scheme
+     */
+    static LookupResult createJSONResult(int status, String body) throws JSONException, IllegalArgumentException {
+        Object json = new JSONTokener(body).nextValue();
+        if (status != 200) {
+            // Immediate consider this is an error for non-200 status code,
+            // try to extract for "error" and "message" in the response body
+            try {
+                return new JSONObjectLookupResult(status, (JSONObject) json);
+            } catch (ClassCastException e) {
+                // If this is not a JSON object then just throw for the caller to handle
+                throw new IllegalArgumentException(e);
+            }
+        } else if (json instanceof JSONObject) {
+            return new JSONObjectLookupResult(status, (JSONObject) json);
+        } else if (json instanceof JSONArray) {
+            return new JSONArrayLookupResult(status, (JSONArray) json);
+        } else {
+            throw new IllegalArgumentException(
+                    "Expect either an JSON Object or Array while received: " +
+                            (json != null ? json.getClass() : "null"));
+        }
+    }
+
+    /**
+     * JSON Array responses are considered success, expect to be an array of Profiles
+     */
+    private static class JSONArrayLookupResult extends JSONLookupResult {
+        private final JSONArray json;
+
+        JSONArrayLookupResult(int responseCode, JSONArray json) {
+            super(responseCode);
+            this.json = json;
+        }
+
+        @Override
+        void invoke(FetchUserSegmentsCallback callback) {
```

Comment:
Should invoke JSONLookupCallback (I'm just suggesting here. Name's up to you). JSONArrayLookupResult should not have anything to do with FetchUserSegmentsCallback

Reply:
I quote the above:
> FetchUserSegmentsCallback just means this is a callback interface for FetchUserSegments operation, nothing more.